FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+  - alignment may be "multi-dimensional", i.e., the `KDTree` example above vs. dimensions aligned independently of each other
+- we need to decide what to do when one dimension has more than one index that supports alignment
+  - we should probably raise unless the user explicitly specify which index to use for the alignment
+- we need to decide what to do when one dimension has one or more index(es) but none support alignment
+  - either we raise or we fail back (silently) to alignment based on dimension size
+- for inexact alignment, the tolerance threshold might be given when building the index and/or when performing the alignment
+- are there cases where we want a specific index to perform alignment and another index to perform selection?
+  - it would be tricky to support that unless we allow multiple indexes per coordinate
+  - alternatively, underlying indexes could be picked internally in a "meta" index for one operation or another, although the risk is to eventually have to deal with an explosion of index wrapper classes with different meta indexes for each combination that we'd like to use.
+
+### 2.7 Using indexes for other purposes
+
+Xarray also provides a number of Dataset/DataArray methods where indexes are used in various ways, e.g.,
+
+- `resample` (`CFTimeIndex` and a `DatetimeIntervalIndex`)
+- `DatetimeAccessor` & `TimedeltaAccessor` properties (`CFTimeIndex` and a `DatetimeIntervalIndex`)
+- `interp` & `interpolate_na`,
+   - with `IntervalIndex`, these become regridding operations. Should we support hooks for these operations?
+- `differentiate`, `integrate`, `polyfit`
+   - raise an error if not a "simple" 1D index?
+- `pad`
+- `coarsen` has to make choices about output index labels.
+- `sortby`
+- `stack`/`unstack`
+- plotting
+    - `plot.pcolormesh` "infers" interval breaks along axes, which are really inferred `bounds` for the appropriate indexes.
+    - `plot.step` again uses `bounds`. In fact, we may even want `step` to be the default 1D plotting function if the axis has `bounds` attached.
+
+It would be reasonable to first restrict those methods to the indexes that are currently available in Xarray, and maybe extend the `XarrayIndex` API later upon request when the opportunity arises.
+
+Conversely, nothing should prevent implementing "non-standard" API in 3rd-party `XarrayIndex` subclasses that could be used in DataArray/Dataset extensions (accessors). For example, we might want to reuse a `KDTree` index to compute k-nearest neighbors (returning a DataArray/Dataset with a new dimension) and/or the distances to the nearest neighbors (returning a DataArray/Dataset with a new data variable).
+
+### 2.8 Index encoding
+
+Indexes don't need to be directly serializable since we could (re)build them from their corresponding coordinate(s). However, we may take advantage that some indexes could be encoded/decoded to/from a set of arrays that would allow optimized reconstruction and/or storage, e.g.,
```

Comment:
```suggestion
Indexes don't need to be directly serializable since we could (re)build them from their corresponding coordinate(s). However, it would be useful if some indexes could be encoded/decoded to/from a set of arrays that would allow optimized reconstruction and/or storage, e.g.,
```